File: VideoListerLibrary/VideoDownloader.cs
/*
 * SPDX-FileCopyrightText: 2021-2022 The Video Lister Contributors
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

using Google.Apis.YouTube.v3;
using Google.Apis.YouTube.v3.Data;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace VideoListerLibrary
{
    /// <summary>
    ///   Asynchronously gets video meta-data using YouTube Data API v3.
    /// </summary>
    public class VideoDownloader
    {
        /// <summary>
        ///   The play-list that is requested to get videos from.
        /// </summary>
        /// <value>
        ///   A long string of characters that represents the play-list that the
        ///   videos belong to.
        /// </value>
        public static string PlayListId { get; set; }

        /// <summary>
        ///   OAuth2 data that is needed to get videos.
        /// </summary>
        /// <value>
        ///   <see cref="AuthCredentials" /> from Google.
        /// </value>
        public static YouTubeService YoutubeService { get; set; }

        /// <summary>
        ///   Asynchronously download 50 video meta-data at a time as a
        ///   play-list item list.
        /// </summary>
        /// <param name="pageToken">
        ///   The token to get the next page. The default value is an empty
        ///   string.
        /// </param>
        /// <returns>
        ///   The meta-data for the 50 videos in the list.
        /// </returns>
        public static async Task<PlaylistItemListResponse> GetPlaylist(string pageToken = "")
        {
            PlaylistItemsResource.ListRequest request =
                YoutubeService.PlaylistItems.List("snippet");
            request.PlaylistId = PlayListId;
            request.MaxResults = 50;
            request.PageToken = pageToken;

            return await request.ExecuteAsync();
        }

        /// <summary>
        ///   Recursively download meta-data from a play-list using
        ///   <see cref="GetPlaylist" />.
        /// </summary>
        /// <returns>
        ///   A list of videos from a play-list.
        /// </returns>
        public static async Task<IList<Video>> GetVideoList()
        {
            IList<Video> videos = new List<Video>();
            string PageToken = "";
            while (PageToken != null)
            {
                PlaylistItemListResponse response = await GetPlaylist(PageToken);
                IList<PlaylistItem> videoItems = response.Items;
                foreach (PlaylistItem item in videoItems)
                {
                    videos.Add(new Video
                    {
                        Title = item.Snippet.Title,
                        Id = item.Snippet.ResourceId.VideoId
                    });
                }
                PageToken = response.NextPageToken;
            }
            return videos;
        }
    }
}

File: Python/video_lister/video_downloader.py
# SPDX-FileCopyrightText: 2022 The Video Lister Contributors

# SPDX-License-Identifier: BSD-3-Clause
"""A method to get video metadata from a playlist."""

from typing import Any

from video_lister import Video


class VideoDownloader:
    """Gets video meta-data using YouTube Data API v3."""

    def __init__(self: "VideoDownloader") -> None:
        self.playlist_id: str = ""
        self.auth_credentials: Any = None

    def get_single_page(self: "VideoDownloader",
                        page_token: str = "") -> Any:
        """Asynchronously download 50 video meta-data at a time as a
        play-list item list.

        Args:
            self (VideoDownloader): The VideoDownloader object to download
            videos.
            page_token (str, optional): _description_. Defaults to an empty
            string.

        Returns:
            Any: The meta-data for the 50 videos in the list.
        """

        return self.auth_credentials.playlistItems().list(
            maxResults=50,
            pageToken=page_token,
            part="snippet",
            playlistId=self.playlist_id
        ).execute()

    def get_video_list(self: "VideoDownloader") -> list[Video]:
        """Recursively download meta-data from a play-list using
        'get_single_page'.

        Args:
            self (VideoDownloader): The VideoDownloader object to download
            videos.

        Returns:
            list[Video]: A list of videos from a play-list.
        """

        videos = []
        page_token = ""

        while page_token is not None:
            response = self.get_single_page(page_token)

            for video in response["items"]:
                videos.append(
                    Video(
                        video["snippet"]["resourceId"]["videoId"],
                        video["snippet"]["title"]
                    )
                )

            page_token = (response["nextPageToken"]
                          if "nextPageToken" in response
                          else None)

        return videos
